pmd-core/src/main/java/net/sourceforge/pmd/reporting/DeterministicOutputListenerWrapper.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.reporting;
+
+import static net.sourceforge.pmd.util.CollectionUtil.listOf;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Objects;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.sourceforge.pmd.lang.document.FileId;
+import net.sourceforge.pmd.lang.document.TextFile;
+import net.sourceforge.pmd.reporting.Report.ConfigurationError;
+import net.sourceforge.pmd.reporting.Report.ProcessingError;
+import net.sourceforge.pmd.reporting.Report.ReportBuilderListener;
+import net.sourceforge.pmd.reporting.Report.SuppressedViolation;
+
+/**
+ * A listener that mediates access to another listener to order events
+ * in a predetermined, stable way. When running PMD with multiple threads,
+ * file listeners may be called in any order. This makes runs non-deterministic,
+ * as Renderers can output reports in the particular order files were processed,
+ * which varies between runs. This class will reorder the events by buffering
+ * them, to call the methods of the underlying GlobalAnalysisListener in a
+ * deterministic order.
+ *
+ * <p>Any renderer whose output may depend on ordering of files should be shielded
+ * by an instance of this wrapper. Note that each wrapper maintains its own buffer
+ * so it should be
+ */
+public class DeterministicOutputListenerWrapper implements GlobalAnalysisListener {
+    private static final Logger LOG = LoggerFactory.getLogger(DeterministicOutputListenerWrapper.class.getName());
+
+    private final GlobalAnalysisListener listener;
+    private final Map<FileId, Integer> filesToIdx = new HashMap<>();
+
+    // use linkedlist because we are mostly doing one-element insertions and removals
+    private final List<ReportWrapper> reportBuffer = new LinkedList<>();
+    private int nextToOutput;
+    private final Object lock = new Object();
```

Comment:
```suggestion
    private final Lock lock = new ReentrantLock();
    private int nextToOutput;
```
* maybe use different lock: as seen here: https://github.com/spring-projects/spring-framework/pull/34538/files
* non final last

Reply:
This might be helpful to do the migration at a glance, and then only have to communicate and sync it properly. Certainly, the CI is not included here, but code-wise, this should be manageable, right? Or are there a lot of breaking changes that aren't covered by the rewrite?

#### Useful Resources:
- [Migrate to Java 21](https://docs.openrewrite.org/running-recipes/popular-recipe-guides/migrate-to-java-21)
- [Migrate to Java 17](https://docs.openrewrite.org/running-recipes/popular-recipe-guides/migrate-to-java-17)

Reply:
If this truly is a Java virtual threads thing, then it should be covered by some kind of recipe, as this seems to be mandatory to migrate to the new concurrency goodness. Java 21 will be a tremendous improvement, and there the virtual threads future will work properly—no more thread pinning issues.

Reply:
> Btw. switching to Java 21 - we could create a multi-release jar and make use of virtual threads, if PMD is run under Java 21. But again - that's a completely different topic...

@adangel

I don't think PMD can benefit from virtual threads. Each parallel task (processing of a file) needs too much memory (to build the AST eg), and the IO waiting is mainly waits from disk loads, which is 1. fast, so there is not much time for context switching and 2. limited in bandwidth, as in, if you open too many files for reading concurrently, you will overload the OS and the memory bus of your system. So there is no way around pooling a small number of threads and limiting the number of concurrent tasks, which is not the programming model of virtual threads. If you profile PMD in a multithreaded run, however, you will find that CPU utilization is very high. We already use all the parallelism we can, so there is also nothing we can gain from virtual threads in terms of parallelism.

Virtual threads were just not designed for applications like ours. They were designed for things like web servers which might be waiting a _really_ long time for a response from a subtask and have nothing to do in between, so you might as well process another request in the meantime. And each waiting subtask requires very little memory, so you can spawn millions if you want. 

In summary, if we wanted to use the virtual threads programming model we would spawn one vthread per file to analyze. But that leads to terrible peak memory usage (as in, program crashes with OOM error), overloads OS resources, and also does not increase effective parallelism.

Reply:
what *could* potentially be beneficial for us is actually splitting away IO operations to a dedicated thread so as to remove all contingency on the disk. Thankfully for us, IO is actually the first step, so we could potentially move to a single thread read a file into memory => submit to a working queue to actually parse and analyze the file.

How much of an impact this would make I'm unsure, specially as SSDs become the de-facto standard.

Reply:
Virtual threads probably won't help much due to the earlier mentioned memory limitations, so their impact may not be as significant as expected.  

The guaranteed benefit is saving native OS resources, which is not a top priority but still a win. This is virtual threads' primary promise besides being seamless to integrate, which turned out to be faulty in the first place. Thread-pinning issues are promised to be resolved in the upcoming Java 24 release. This was not a surprise for such a high-tier topic.  

Virtual threads help avoid naive OS thread usage by pooling virtual threads into actual OS threads. In some edge cases on small systems, this might provide speed improvements since, without virtual threads, processes would just wait when no native thread is available. Virtual threads could potentially release resources for other uses, though this is highly speculative.